Networking/Networking/Model/Product/Product.swift

```diff
@@ -0,0 +1,534 @@
+import Foundation
+
+
+/// Represents a Product Entity.
+///
+public struct Product: Decodable {
+    public let siteID: Int
+    public let productID: Int
+    public let name: String
+    public let slug: String
+    public let permalink: String
+
+    public let dateCreated: Date        // gmt
+    public let dateModified: Date?      // gmt
+
+    public let productTypeKey: String
+    public let statusKey: String        // draft, pending, private, published
+    public let featured: Bool
+    public let catalogVisibilityKey: String // visible, catalog, search, hidden
+
+    public let description: String?
+    public let shortDescription: String?
+    public let sku: String?
+
+    public let price: String
+    public let regularPrice: String?
+    public let salePrice: String?
+    public let onSale: Bool
+
+    public let purchasable: Bool
+    public let totalSales: Int
+    public let virtual: Bool
+
+    public let downloadable: Bool
+    public let downloadLimit: Int       // defaults to -1
+    public let downloadExpiry: Int      // defaults to -1
+
+    public let externalURL: String?
+    public let taxStatusKey: String     // taxable, shipping, none
+    public let taxClass: String?
+
+    public let manageStock: Bool
+    public let stockQuantity: Int?      // API reports Int or null
+    public let stockStatusKey: String   // instock, outofstock, backorder
+
+    public let backordersKey: String    // no, notify, yes
+    public let backordersAllowed: Bool
+    public let backordered: Bool
+
+    public let soldIndividually: Bool
+    public let weight: String?
+    public let dimensions: Dimensions
+
+    public let shippingRequired: Bool
+    public let shippingTaxable: Bool
+    public let shippingClass: String?
+    public let shippingClassID: Int
+
+    public let reviewsAllowed: Bool
+    public let averageRating: String
+    public let ratingCount: Int
+
+    public let relatedIDs: [Int?]
+    public let upsellIDs: [Int?]
+    public let crossSellIDs: [Int?]
+    public let parentID: Int
+
+    public let purchaseNote: String?
+    public let categories: [ProductCategory?]
+    public let tags: [ProductTag?]
+    public let images: [ProductImage?]
+
+    public let attributes: [ProductAttribute?]
+    public let defaultAttributes: [ProductDefaultAttribute?]
+    public let variations: [Int?]
+    public let groupedProducts: [Int?]
```

Comment:
Re: all of these collections of related models (`categories`, `tags`, `images`, `attributes`, `defaultAttributes`, `variations`, `groupedProducts`), do they need to be arrays of optional, or could they simply be an array of the non-option structs? (e.g. `tags: [ProductTag]`)

I'm thinking of how we handled this elsewhere in the app — [order coupons](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Model/Order.swift#L32) is a good example.  Also if it's possible that a given property could be nil, we could always use the coalescing operator [like we did in notifications ](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Model/Note.swift#L166)(and elsewhere) when parsing the JSON.